File: Source/Plugins/Workflow/com.tle.web.workflow/src/com/tle/web/workflow/tasks/dialog/AbstractTaskActionDialog.java
package com.tle.web.workflow.tasks.dialog;

import com.google.common.base.Throwables;
import com.google.common.collect.Lists;
import com.tle.annotation.NonNullByDefault;
import com.tle.annotation.Nullable;
import com.tle.common.Check;
import com.tle.common.filesystem.FileEntry;
import com.tle.common.filesystem.handle.StagingFile;
import com.tle.core.filesystem.staging.service.StagingService;
import com.tle.core.institution.InstitutionService;
import com.tle.core.services.FileSystemService;
import com.tle.web.freemarker.FreemarkerFactory;
import com.tle.web.freemarker.annotations.ViewFactory;
import com.tle.web.resources.PluginResourceHelper;
import com.tle.web.resources.ResourcesService;
import com.tle.web.sections.SectionInfo;
import com.tle.web.sections.SectionTree;
import com.tle.web.sections.ajax.AjaxGenerator;
import com.tle.web.sections.ajax.handler.AjaxMethod;
import com.tle.web.sections.ajax.handler.UpdateDomFunction;
import com.tle.web.sections.annotations.Bookmarked;
import com.tle.web.sections.annotations.EventHandlerMethod;
import com.tle.web.sections.equella.annotation.PlugKey;
import com.tle.web.sections.equella.dialog.AbstractOkayableDialog;
import com.tle.web.sections.equella.render.ButtonRenderer;
import com.tle.web.sections.events.ReadyToRespondListener;
import com.tle.web.sections.events.RenderContext;
import com.tle.web.sections.events.js.BookmarkAndModify;
import com.tle.web.sections.jquery.libraries.JQueryProgression;
import com.tle.web.sections.js.JSAssignable;
import com.tle.web.sections.js.JSCallAndReference;
import com.tle.web.sections.js.JSCallable;
import com.tle.web.sections.js.JSExpression;
import com.tle.web.sections.js.generic.Js;
import com.tle.web.sections.js.generic.OverrideHandler;
import com.tle.web.sections.js.generic.StatementHandler;
import com.tle.web.sections.js.generic.function.*;
import com.tle.web.sections.render.Label;
import com.tle.web.sections.render.SectionRenderable;
import com.tle.web.sections.render.TextLabel;
import com.tle.web.sections.standard.Button;
import com.tle.web.sections.standard.FileDrop;
import com.tle.web.sections.standard.TextField;
import com.tle.web.sections.standard.annotations.Component;
import com.tle.web.sections.standard.dialog.model.DialogModel;
import com.tle.web.sections.standard.model.HtmlLinkState;
import com.tle.web.sections.standard.model.SimpleBookmark;
import com.tle.web.workflow.servlet.WorkflowMessageServlet;
import com.tle.web.workflow.tasks.CurrentTaskSection;

import javax.inject.Inject;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@NonNullByDefault
public abstract class AbstractTaskActionDialog<M extends AbstractTaskActionDialog.AbstractTaskActionDialogModel> extends AbstractOkayableDialog<M> implements ReadyToRespondListener
{
	private static final PluginResourceHelper URL_HELPER = ResourcesService.getResourceHelper(AbstractTaskActionDialog.class);
	private static final IncludeFile INCLUDE = new IncludeFile(URL_HELPER.url("scripts/comments.js"));
	private static final JSCallAndReference WORKFLOW_COMMENTS_CLASS = new ExternallyDefinedFunction("WorkflowComments",
			INCLUDE, JQueryProgression.PRERENDER);
	private static final ExternallyDefinedFunction VALIDATE_FILE = new ExternallyDefinedFunction(WORKFLOW_COMMENTS_CLASS,
			"validateFile", 2);

	private JSAssignable validateFile;

	@Inject
	protected FileSystemService fileSystemService;
	@Inject
	protected StagingService stagingService;

	@PlugKey("command.taskaction.attachedfiles")
	private static Label LABEL_ATTACHED_FILES;
	@PlugKey("command.taskaction.mustmessage")
	private static Label LABEL_MUST_MESSAGE;

	@Component
	private TextField commentField;
	@Component(name = "fd")
	@PlugKey("comments.upload")
	private FileDrop fileDrop;
	@Component(name = "u")
	@PlugKey("command.taskaction.uploadfile")
	private Button uploadButton;
	@Component(name = "c")
	@PlugKey("command.taskaction.cancel")
	private Button cancelButton;

	@ViewFactory
	protected FreemarkerFactory viewFactory;
	private JSCallable successCallable;
	private UpdateDomFunction removeCallback;

	protected AbstractTaskActionDialog()
	{
		setAjax(true);
	}

	protected abstract ButtonRenderer.ButtonType getButtonType();

	protected abstract Label getButtonLabel();

	protected abstract CurrentTaskSection.CommentType getActionType();

	public abstract Label getPostCommentHeading();

	@Nullable
	protected String getWorkflowStepTarget(SectionInfo info)
	{
		return null;
	}

	@Override
	public String getWidth()
	{
		return "696px";
	}

	@Override
	protected SectionRenderable getRenderableContents(RenderContext context)
	{
		setupModelForRender(context);
		return viewFactory.createResult("dialog/taskaction.ftl", this);
	}

	@Override
	protected String getContentBodyClass(RenderContext context)
	{
		return "taskactiondialog";
	}

	@Override
	public void readyToRespond(SectionInfo info, boolean redirect)
	{
		super.readyToRespond(info, redirect);
		M model = getModel(info);
		if (model.isShowing())
		{
			String stagingFolderUuid = model.getStagingFolderUuid();
			if (stagingFolderUuid == null)
			{
				stagingFolderUuid = stagingService.createStagingArea().getUuid();
				model.setStagingFolderUuid(stagingFolderUuid);
			}
		}
	}


	protected void setupModelForRender(RenderContext context)
	{
		M model = getModel(context);
		String stagingFolderUuid = model.getStagingFolderUuid();

		try
		{
			FileEntry[] fileEntries = fileSystemService.enumerate(new StagingFile(stagingFolderUuid), null, null);
			ArrayList<HtmlLinkState> lists = Lists.newArrayList();
			ArrayList<HtmlLinkState> deleteFiles = Lists.newArrayList();
			int index = 0;
			for( FileEntry fileEntry : fileEntries )
			{
				String filename = fileEntry.getName();
				HtmlLinkState link = new HtmlLinkState(new TextLabel(filename),
						new SimpleBookmark(WorkflowMessageServlet.stagingUrl(stagingFolderUuid, filename)));
				link.setTarget("_blank");
				lists.add(link);
				HtmlLinkState removeBtn = new HtmlLinkState(new TextLabel(""));
				removeBtn.addClass("unselect");
				removeBtn.setClickHandler(new OverrideHandler(removeCallback, filename));
				deleteFiles.add(removeBtn);
				index++;
			}
			model.setStagingFiles(lists);
			model.setDeleteFiles(deleteFiles);
		}
		catch( IOException e )
		{
			throw Throwables.propagate(e);
		}

		fileDrop.setAjaxUploadUrl(context, new BookmarkAndModify(context, ajaxEvents.getModifier("dndUpload")));
		fileDrop.setValidateFile(context, validateFile);
	}

//	@EventHandlerMethod
//	public void ok(SectionInfo info, String comment, String stagingUuid, String workflowStep)
//	{
//		doSaveAction(info, comment, stagingUuid, workflowStep);
//	}
//
//	protected void doSaveAction(SectionInfo info, String comment, String stagingUuid, String workflowStep)
//	{
//		StagingFile stagingFolder = new StagingFile(stagingUuid);
//		long countFiles = fileSystemService.countFiles(stagingFolder, null);
//
//		// when approving with attached files, comment msg cannot be empty
//		/*
//		boolean acceptedWithFiles = Check.isEmpty(comment) && countFiles > 1;
//		if( acceptedWithFiles )
//		{
//			model.setErrorMessage(LABEL_ENTERMSG_WITHFILES);
//			info.preventGET();
//			return;
//		}*/
//
//		CurrentTaskSection currentTaskSection = info.lookupSection(CurrentTaskSection.class);
//		if( !currentTaskSection.doComment(info, getActionType(), workflowStep, comment, stagingUuid) )
//		{
//			//commentField.setValue(info, null);
//			closeDialog(info, new ReloadHandler());
//		}
//	}

	@EventHandlerMethod
	public void ok(SectionInfo info)
	{
		M model = getModel(info);
		String stagingUuid = model.getStagingFolderUuid();
		Label errorMessage = validate(info);
		if (errorMessage == null)
		{
			CurrentTaskSection currentTaskSection = info.lookupSection(CurrentTaskSection.class);
			currentTaskSection.doComment(info, getActionType(), getWorkflowStepTarget(info), commentField.getValue(info), stagingUuid);
			closeDialog(info, successCallable, getActionType());
		}
		else
		{
			info.preventGET();
			model.setErrorMessage(errorMessage);
		}

	}

	protected abstract Label validate(SectionInfo info);

	protected Label validateHasMessage(SectionInfo info)
	{
		if (Check.isEmpty(commentField.getValue(info).trim()))
		{
			return LABEL_MUST_MESSAGE;
		}
		return null;
	}

	@EventHandlerMethod
	public void cancelled(SectionInfo info)
	{
		StagingFile temp = new StagingFile(getModel(info).getStagingFolderUuid());
		stagingService.removeStagingArea(temp, true);
		closeDialog(info);
	}

	@AjaxMethod
	public DndUploadResponse dndUpload(SectionInfo info) throws IOException
	{
		M model = getModel(info);
		String stagingFolderUuid = model.getStagingFolderUuid();
		String filename = fileDrop.getFilename(info);

		String fn = "";
		if( !Check.isEmpty(filename) )
		{
			fn = filename.toLowerCase();
		}
		InputStream stream = fileDrop.getInputStream(info);

		StagingFile staging = new StagingFile(stagingFolderUuid);

		// overwrite existing file
		if( fileSystemService.fileExists(staging, fn) )
		{
			fileSystemService.removeFile(staging, fn);
		}
		fileSystemService.write(staging, fn, stream, false);
		String url = WorkflowMessageServlet.stagingUrl(stagingFolderUuid, fn);
		return new DndUploadResponse(stagingFolderUuid, url);
	}

	@EventHandlerMethod
	public void removeUploadedFile(SectionInfo info, String filename) throws IOException
	{
		StagingFile staging = new StagingFile(getModel(info).getStagingFolderUuid());
		fileSystemService.removeFile(staging, filename);
	}

	@Override
	protected Label getOkLabel()
	{
		return getButtonLabel();
	}

	@Override
	public void treeFinished(String id, SectionTree tree)
	{
		setOkHandler(events.getSubmitValuesHandler("ok"));
		setCancelHandler(new StatementHandler(Js.call_s(new RuntimeFunction()
		{
			@Override
			protected JSCallable createFunction(RenderContext info)
			{
				return new PassThroughFunction("cl", events.getSubmitValuesFunction("cancelled"));
			}
		})));

		cancelButton.setClickHandler(events.getSubmitValuesHandler("cancelled"));
		UpdateDomFunction updateProgressArea = ajaxEvents.getAjaxUpdateDomFunction(tree, this, null,
				ajaxEvents.getEffectFunction(AjaxGenerator.EffectType.REPLACE_IN_PLACE), "uploaded");
		removeCallback = ajaxEvents.getAjaxUpdateDomFunction(tree, this, events.getEventHandler("removeUploadedFile"),
				ajaxEvents.getEffectFunction(AjaxGenerator.EffectType.REPLACE_IN_PLACE), "uploaded");
		validateFile = Js.functionValue(Js.call(VALIDATE_FILE, PartiallyApply.partial(updateProgressArea, 0)));

		super.treeFinished(id, tree);
	}

	@Override
	public M instantiateDialogModel(SectionInfo info)
	{
		return (M)new AbstractTaskActionDialogModel();
	}

	@Override
	protected Collection<Button> collectFooterActions(RenderContext context)
	{
		List<Button> buttons = new ArrayList<>();
		buttons.add(getOk());
		buttons.add(cancelButton);
		return buttons;
	}

	public void setSuccessCallable(JSCallable successCallable)
	{
		this.successCallable = addParentCallable(successCallable);
	}

	public static class DndUploadResponse
	{
		public String stagingUuid;
		public String stagingFileUrl;

		public DndUploadResponse(String stagingUuid, String stagingFileUrl)
		{
			this.stagingUuid = stagingUuid;
			this.stagingFileUrl = stagingFileUrl;
		}
	}

	public static class AbstractTaskActionDialogModel extends DialogModel
	{
		@Bookmarked
		private String stagingFolderUuid;
		private ArrayList<HtmlLinkState> stagingFiles;
		private ArrayList<HtmlLinkState> deleteFiles;
		private Label errorMessage;

		public String getStagingFolderUuid()
		{
			return stagingFolderUuid;
		}

		public void setStagingFolderUuid(String stagingFolderUuid)
		{
			this.stagingFolderUuid = stagingFolderUuid;
		}

		public ArrayList<HtmlLinkState> getStagingFiles()
		{
			return stagingFiles;
		}

		public void setStagingFiles(ArrayList<HtmlLinkState> stagingFiles)
		{
			this.stagingFiles = stagingFiles;
		}

		public ArrayList<HtmlLinkState> getDeleteFiles()
		{
			return deleteFiles;
		}

		public void setDeleteFiles(ArrayList<HtmlLinkState> deleteFiles)
		{
			this.deleteFiles = deleteFiles;
		}

		public Label getErrorMessage()
		{
			return errorMessage;
		}

		public void setErrorMessage(Label errorMessage)
		{
			this.errorMessage = errorMessage;
		}
	}

	public TextField getCommentField()
	{
		return commentField;
	}

	public FileDrop getFileDrop()
	{
		return fileDrop;
	}

	public Button getUploadButton()
	{
		return uploadButton;
	}

	public Label getAttachedFilesLabel()
	{
		return LABEL_ATTACHED_FILES;
	}
}

File: Source/Plugins/Workflow/com.tle.web.workflow/scalasrc/com/tle/web/workflow/tasks/dialog/AbstractTaskActionDialog.scala
/*
 * Copyright 2017 Apereo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tle.web.workflow.tasks.dialog

import java.util
import java.util.{Collections, UUID}
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicReference
import javax.inject.Inject

import com.dytech.edge.exceptions.BannedFileException
import com.tle.common.Check
import com.tle.common.filesystem.handle.StagingFile
import com.tle.core.filesystem.staging.service.StagingService
import com.tle.core.services.FileSystemService
import com.tle.web.freemarker.FreemarkerFactory
import com.tle.web.freemarker.annotations.ViewFactory
import com.tle.web.resources.ResourcesService
import com.tle.web.sections.ajax.AjaxGenerator
import com.tle.web.sections.ajax.handler.{AjaxMethod, UpdateDomFunction}
import com.tle.web.sections.annotations.{Bookmarked, EventHandlerMethod}
import com.tle.web.sections.equella.ajaxupload.AjaxUpload
import com.tle.web.sections.equella.annotation.PlugKey
import com.tle.web.sections.equella.dialog.AbstractOkayableDialog
import com.tle.web.sections.equella.render.{ButtonRenderer, UnselectLinkRenderer}
import com.tle.web.sections.events.js.BookmarkAndModify
import com.tle.web.sections.events.{ReadyToRespondListener, RenderContext}
import com.tle.web.sections.jquery.JQuerySelector
import com.tle.web.sections.js.generic.function.{PartiallyApply, PassThroughFunction, RuntimeFunction}
import com.tle.web.sections.js.generic.{Js, StatementHandler}
import com.tle.web.sections.js.{JSAssignable, JSCallable}
import com.tle.web.sections.render.{Label, SectionRenderable, TextLabel}
import com.tle.web.sections.result.util.KeyLabel
import com.tle.web.sections.standard.annotations.Component
import com.tle.web.sections.standard.dialog.model.DialogModel
import com.tle.web.sections.standard.model.{HtmlComponentState, HtmlLinkState, SimpleBookmark}
import com.tle.web.sections.standard.renderers.DivRenderer
import com.tle.web.sections.standard.{Button, FileDrop, TextField}
import com.tle.web.sections.{SectionInfo, SectionTree}
import com.tle.web.workflow.servlet.WorkflowMessageServlet
import com.tle.web.workflow.tasks.CurrentTaskSection

import scala.collection.JavaConverters._
import scala.util.{Failure, Success}

case class UploadedFile(getLink: HtmlLinkState, getCancel: UnselectLinkRenderer, getProgressDiv: SectionRenderable)

class TaskActionDialogModel extends DialogModel {
  @Bookmarked var stagingFolderUuid: String = null
  var getStagingFiles: util.List[UploadedFile] = null
  var getErrorMessage: Label = null
}

object AbstractTaskActionDialog
{
  val r = ResourcesService.getResourceHelper(getClass)
  def l(key: String) = new KeyLabel(r.key(key))
  def error(k: String) = l("command.taskaction.error."+k)
  val LABEL_ATTACHED_FILES = l("command.taskaction.attachedfiles")
  val LABEL_MUST_MESSAGE = l("command.taskaction.mustmessage")
  val MAX_FILESIZE = 1024 * 1024 * 1024
}

abstract class AbstractTaskActionDialog extends AbstractOkayableDialog[TaskActionDialogModel] with ReadyToRespondListener {
  import AbstractTaskActionDialog._

  setAjax(true)
  var successCallable : JSCallable = _
  var removeCallback : UpdateDomFunction = _
  var cancelFunc : UpdateDomFunction = _
  var validateFile : JSAssignable = _

  @Inject protected var fileSystemService: FileSystemService = _
  @Inject protected var stagingService: StagingService = _

  @Component(name="commentField") var _commentField : TextField = _
  @Component(name = "fd")
  @PlugKey("comments.upload") var _fileDrop : FileDrop = _
  @Component(name = "c")
  @PlugKey("command.taskaction.cancel") var cancelButton : Button = _

  @ViewFactory protected var viewFactory: FreemarkerFactory = _

  override def instantiateDialogModel(info: SectionInfo) = new TaskActionDialogModel

  def getButtonType: ButtonRenderer.ButtonType

  def getButtonLabel: Label

  def getActionType: CurrentTaskSection.CommentType

  def getPostCommentHeading: Label

  private val currentUploads = new ConcurrentHashMap[UUID, (String, AtomicReference[Boolean], Option[Label])]().asScala

  protected def getWorkflowStepTarget(info: SectionInfo): String = null

  override def getWidth = "696px"

  protected def setupModelForRender(context: RenderContext): Unit = {
    val model = getModel(context)
    val stagingFolderUuid = model.stagingFolderUuid
    val fileEntries = fileSystemService.enumerate(new StagingFile(stagingFolderUuid), null, null)
    val files = fileEntries.zipWithIndex.flatMap {
      case (fileEntry, index) =>
        val filename = fileEntry.getName
        val shouldShow = currentUploads.find(_._2._1 == filename) match {
          case None => Some((index.toString, true, new StatementHandler(removeCallback, filename)))
          case Some((uuid, (fn, ar, err))) if !ar.get() => Some((uuid, false, new StatementHandler(cancelFunc, uuid)))
          case _ => None
        }
        shouldShow.map { case (id, finished, removeHandler) =>
          val link = new HtmlLinkState(new TextLabel(filename), new SimpleBookmark(WorkflowMessageServlet.stagingUrl(stagingFolderUuid, filename)))
          link.setTarget("_blank")
          val remove = new HtmlLinkState
          remove.setClickHandler(removeHandler)
          val cancel = new UnselectLinkRenderer(remove, new TextLabel(""))
          val progressDivState = new HtmlComponentState
          val progressDiv = new DivRenderer(progressDivState)
          progressDivState.setId("u" + id)
          if (finished) {
            val inner = new DivRenderer("")
            inner.addClass("progress-bar-inner complete")
            progressDiv.setNestedRenderable(inner)
          }
          UploadedFile(link, cancel, progressDiv)
        }
    }
    currentUploads --= currentUploads.collect {
      case (u, (_,_, Some(m))) =>
        model.getErrorMessage = m
        u
    }
    model.getStagingFiles = files.toList.asJava
    _fileDrop.setAjaxUploadUrl(context, new BookmarkAndModify(context, ajaxEvents.getModifier("dndUpload")))
    _fileDrop.setValidateFile(context, validateFile)
  }

  override protected def getRenderableContents(context: RenderContext): SectionRenderable = {
    setupModelForRender(context)
    viewFactory.createResult("dialog/taskaction.ftl", this)
  }

  override protected def getContentBodyClass(context: RenderContext) = "taskactiondialog"

  override def readyToRespond(info: SectionInfo, redirect: Boolean): Unit = {
    super.readyToRespond(info, redirect)
    val model = getModel(info)
    if (model.isShowing) {
      var stagingFolderUuid = model.stagingFolderUuid
      if (stagingFolderUuid == null) {
        stagingFolderUuid = stagingService.createStagingArea.getUuid
        model.stagingFolderUuid = stagingFolderUuid
      }
    }
  }

  def validate(info: SectionInfo): Label

  def isMandatoryMessage : Boolean

  def validateHasMessage(info: SectionInfo): Label = {
    if (Check.isEmpty(_commentField.getValue(info).trim()))
    {
      AbstractTaskActionDialog.LABEL_MUST_MESSAGE
    } else null
  }

  def setSuccessCallable(successCallable : JSCallable) = this.successCallable = addParentCallable(successCallable)


  @EventHandlerMethod def ok(info: SectionInfo): Unit = {
    val model = getModel(info)
    val stagingUuid = model.stagingFolderUuid
    val errorMessage = validate(info)
    if (errorMessage == null) {
      val currentTaskSection : CurrentTaskSection = info.lookupSection(classOf[CurrentTaskSection])
      currentTaskSection.doComment(info, getActionType, getWorkflowStepTarget(info), _commentField.getValue(info), stagingUuid)
      closeDialog(info, successCallable, getActionType)
    }
    else {
      info.preventGET()
      model.getErrorMessage = errorMessage
    }
  }

  @EventHandlerMethod def cancelled(info: SectionInfo): Unit = {
    val temp = new StagingFile(getModel(info).stagingFolderUuid)
    stagingService.removeStagingArea(temp, true)
    closeDialog(info)
  }

  override def getOkLabel = getButtonLabel
  def getAttachedFilesLabel: Label = LABEL_ATTACHED_FILES

  def getCommentField : TextField = _commentField
  def getFileDrop : FileDrop = _fileDrop

  @AjaxMethod
  def dndUpload(info: SectionInfo): Boolean = {
    val model = getModel(info)
    val uploadId = UUID.fromString(info.getRequest.getHeader("X_UUID"))
    val stagingFolderUuid = model.stagingFolderUuid
    val filename = _fileDrop.getFilename(info)
    val fn = if (!Check.isEmpty(filename)) filename.toLowerCase else ""
    val stream = _fileDrop.getInputStream(info)
    val staging = new StagingFile(stagingFolderUuid)
    val uploadRef = new AtomicReference[Boolean](false)
    currentUploads.update(uploadId, (fn, uploadRef, None))
    // overwrite existing file
    if (fileSystemService.fileExists(staging, fn))
      fileSystemService.removeFile(staging, fn)
    AjaxUpload.writeCancellableStream(s => fileSystemService.write(staging, fn, s, false), stream, uploadRef) match {
      case Success(_) => currentUploads.remove(uploadId)
      case Failure(f) => fileSystemService.removeFile(staging, fn)
        f match {
          case bfe: BannedFileException => currentUploads.put(uploadId, (fn, uploadRef, Some(error("banned"))))
          case _ => currentUploads.remove(uploadId)
        }
    }
    true
  }

  @EventHandlerMethod
  def removeUploadedFile(info: SectionInfo, filename: String): Unit = {
    val staging = new StagingFile(getModel(info).stagingFolderUuid)
    fileSystemService.removeFile(staging, filename)
  }

  @EventHandlerMethod
  def illegalFile(info: SectionInfo, filename: String, reason: String): Unit = {
    getModel(info).getErrorMessage = error(reason)
  }

  @EventHandlerMethod
  def cancelFile(info: SectionInfo, uuid: UUID): Unit = {
    currentUploads.get(uuid).foreach {
      case (fn, ar, _) => ar.set(true)
    }

  }



  override def treeFinished(id: String, tree: SectionTree): Unit = {
    setOkHandler(events.getSubmitValuesHandler("ok"))
    setCancelHandler(new StatementHandler(Js.call_s(new RuntimeFunction() {
      override protected def createFunction(info: RenderContext) = new PassThroughFunction("cl", events.getSubmitValuesFunction("cancelled"))
    })))
    cancelButton.setClickHandler(events.getSubmitValuesHandler("cancelled"))
    val divs = Seq("uploads", "comment-error")
    val updateProgressArea = ajaxEvents.getAjaxUpdateDomFunction(tree, this, null,
      ajaxEvents.getEffectFunction(AjaxGenerator.EffectType.REPLACE_IN_PLACE), divs: _*)

    removeCallback = ajaxEvents.getAjaxUpdateDomFunction(tree, this,
      events.getEventHandler("removeUploadedFile"), ajaxEvents.getEffectFunction(AjaxGenerator.EffectType.REPLACE_IN_PLACE), divs: _*)

    cancelFunc = ajaxEvents.getAjaxUpdateDomFunction(tree, this, events.getEventHandler("cancelFile"),
      ajaxEvents.getEffectFunction(AjaxGenerator.EffectType.REPLACE_IN_PLACE), divs: _*)

    val errorCallback = PartiallyApply.partial(events.getSubmitValuesFunction("illegalFile"), 2)
    val startedUpload = AjaxUpload.createProgressFunc(new JQuerySelector(JQuerySelector.Type.RAW, "#uploads .uploadsprogress"), cancelFunc)
    validateFile = Js.functionValue(AjaxUpload.createValidate(MAX_FILESIZE, Collections.emptyList[String],
      errorCallback, startedUpload, PartiallyApply.partial(updateProgressArea, 0)))
    super.treeFinished(id, tree)
  }

  override protected def collectFooterActions(context: RenderContext): util.Collection[Button] = {
    val buttons = new util.ArrayList[Button]
    buttons.add(getOk)
    buttons.add(cancelButton)
    buttons
  }
}
